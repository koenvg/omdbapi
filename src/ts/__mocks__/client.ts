
export const get = (params: any): Promise<any> => {
  return  new Promise((resolve, reject) => {
    resolve();
  });
};
