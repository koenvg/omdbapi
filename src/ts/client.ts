import { settings } from "./config/config";
import "whatwg-fetch";

const paramsMap = {
  search: "s",
  title: "t",
  imdbId: "i",
  year: "y",
  page: "page",
  season: "Season",
  episode: "Episode",
  plot: "plot",
  type: "type",
  tomatoes: "tomatoes",
};

const createUrlFromParams = (params: any): string => {
  const keys = Object.keys(params);
  let url = `${settings.url}?r=json`;
  for (let i = 0; i < keys.length; i++) {
    const key = keys[i];
    const convertedKey = paramsMap[key];
    if (!convertedKey) {
      throw new Error(`Invalid key ${key}`);
    }
    url += `&${convertedKey}=${params[key]}`;
  }
  return url;
};

export const get = (params: any): Promise<any> => {
  const url = createUrlFromParams(params);
  return fetch(url)
    .then((res) => {
      return res.json();
    })
    .then((searchResult: any) => {
      if (searchResult.Response === "False") {
        throw Error(searchResult.Error);
      }
      return searchResult;
    });
};
