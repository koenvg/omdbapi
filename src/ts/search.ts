import { get } from "./client";

export type TypeOfContent = "movie" | "series" | "episode";

export interface SearchResult {
  Title: string;
  Year: string;
  imdbID: string;
  Type: TypeOfContent;
  Poster: string;
  Released?: string;
  Runtime?: string;
  Genre?: string;
  Director?: string;
  Writer?: string;
  Actors?: string;
  Plot?: string;
  Language?: string;
  Country?: string;
  Awards?: string;
  Metascrore?: string;
  imdbRating?: string;
  imdbVotes?: string;
  totalSeasons?: number;
}

export interface SeasonResult {
  season: number;
  episodes: SearchResult[];
};

export interface ShowResult extends SearchResult {
  seasons: SeasonResult[];
}

export interface ApiResult {
  Search: SearchResult[];
  totalResults: string;
  Response: string;
}

export interface RequestParams {
  contentType?: TypeOfContent;
  year?: string;
  page?: number;
  plot?: string;
  includeTomatoesRating?: boolean;
}

export const findByIdWithSeasons = (imdbId: string): Promise<ShowResult> => {
  let show: SearchResult;
  return get({ imdbId })
    .then(result => {
      show = result;
      const seasons = [];
      for (let i = 1; i <= result.totalSeasons; i++) {
        seasons.push(get({ imdbId, season: i })
          .then(season => ({ ...season, episodes: season.Episodes })));
      }
      return Promise.all(seasons);
    })
    .then((seasons: SeasonResult[]) => {
      return { ...show, seasons };
    });
};

export const findById = (imdbId: string, searchParams?: RequestParams): Promise<SearchResult> => {
  const params = { ...searchParams, imdbId };
  return get(params);
};

export const findByTitle = (title: string, searchParams: RequestParams = {}): Promise<SearchResult> => {
  const params = { ...searchParams, title };
  return get(params);
};

export const search = (query: string, searchParams?: RequestParams): Promise<ApiResult> => {
  const params = { ...searchParams, search: query };
  return get(params);
};
