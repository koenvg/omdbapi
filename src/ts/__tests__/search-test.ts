jest.dontMock("../search");
jest.mock("../client");

import { findByTitle, findById, search, RequestParams, findByIdWithSeasons } from "../search";
import * as Client from "../client";

describe("findByTitle", () => {
  const params: RequestParams = {
    includeTomatoesRating: false,
  };
  let get: jasmine.Spy;

  beforeEach(() => {
    get = spyOn(Client, "get").and.callFake(() => {
      return new Promise<void>((resolve, reject) => {
        resolve();
      });
    });
  });

  it("should add title to the params", () => {
    findByTitle("title");
    expect(get).toBeCalledWith({
      title: "title",
    });
  });

  it("should pass on the params", () => {
    findByTitle("title", params);
    expect(get).toBeCalledWith({
      title: "title",
      includeTomatoesRating: false,
    });
  });
});

describe("findById", () => {
  const params: RequestParams = {
    includeTomatoesRating: false,
  };
  let get: jasmine.Spy;

  beforeEach(() => {
    get = spyOn(Client, "get").and.callFake(() => {
      return new Promise<void>((resolve, reject) => {
        resolve();
      });
    });
  });

  it("should add imdbId to the params", () => {
    findById("id");
    expect(get).toBeCalledWith({
      imdbId: "id",
    });
  });

  it("should pass on the params", () => {
    findById("id", params);
    expect(get).toBeCalledWith({
      imdbId: "id",
      includeTomatoesRating: false,
    });
  });
});

describe("search", () => {
  const params: RequestParams = {
    includeTomatoesRating: false,
  };
  let get: jasmine.Spy;

  beforeEach(() => {
    get = spyOn(Client, "get").and.callFake(() => {
      return new Promise<void>((resolve, reject) => {
        resolve();
      });
    });
  });

  it("should add search to the params", () => {
    search("search");
    expect(get).toBeCalledWith({
      search: "search",
    });
  });

  it("should pass on the params", () => {
    search("search", params);
    expect(get).toBeCalledWith({
      search: "search",
      includeTomatoesRating: false,
    });
  });
});

describe("findByIdWithSeasons", () => {
  (Client as any).get = jest.fn().mockReturnValue(Promise.resolve({ title: "Hello world", totalSeasons: "12" }));

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("should return the show details with the season details", async () => {
    const show = await findByIdWithSeasons("id");
    expect(show).toMatchSnapshot();
  });

  it("should make a call for each season", async () => {
    await findByIdWithSeasons("id");
    for (let i = 1; i <= 12; i++) {
      expect(Client.get).toBeCalledWith({ imdbId: "id", season: i });
    }
  });

});
