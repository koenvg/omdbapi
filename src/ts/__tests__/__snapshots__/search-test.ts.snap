// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`findByIdWithSeasons should return the show details with the season details 1`] = `
Object {
  "seasons": Array [
    Object {
      "episodes": undefined,
      "title": "Hello world",
      "totalSeasons": "12",
    },
    Object {
      "episodes": undefined,
      "title": "Hello world",
      "totalSeasons": "12",
    },
    Object {
      "episodes": undefined,
      "title": "Hello world",
      "totalSeasons": "12",
    },
    Object {
      "episodes": undefined,
      "title": "Hello world",
      "totalSeasons": "12",
    },
    Object {
      "episodes": undefined,
      "title": "Hello world",
      "totalSeasons": "12",
    },
    Object {
      "episodes": undefined,
      "title": "Hello world",
      "totalSeasons": "12",
    },
    Object {
      "episodes": undefined,
      "title": "Hello world",
      "totalSeasons": "12",
    },
    Object {
      "episodes": undefined,
      "title": "Hello world",
      "totalSeasons": "12",
    },
    Object {
      "episodes": undefined,
      "title": "Hello world",
      "totalSeasons": "12",
    },
    Object {
      "episodes": undefined,
      "title": "Hello world",
      "totalSeasons": "12",
    },
    Object {
      "episodes": undefined,
      "title": "Hello world",
      "totalSeasons": "12",
    },
    Object {
      "episodes": undefined,
      "title": "Hello world",
      "totalSeasons": "12",
    },
  ],
  "title": "Hello world",
  "totalSeasons": "12",
}
`;
